crypto/s2n_locking.c

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include <openssl/crypto.h>
+#include <pthread.h>
+
+#include "crypto/s2n_locking.h"
+#include "crypto/s2n_openssl.h"
+#include "utils/s2n_mem.h"
+#include "utils/s2n_safety.h"
+
+/* Writing multithreaded applications using Openssl-1.0.2
+ * requires calling CRYPTO_set_locking_callback.
+ * If the callback is not set, locks are no-ops and unexpected
+ * behavior may occur, particularly for RSA and X509.
+ *
+ * In the past s2n-tls relied on customers setting the callback
+ * themselves, but that seems unnecessary since other parts of
+ * the library (like fork detection) already rely on the pthreads library.
+ *
+ * For more information:
+ * https://www.openssl.org/blog/blog/2017/02/21/threads/
+ * https://www.openssl.org/docs/man1.0.2/man3/threads.html
+ */
+
+#define S2N_MUTEXES(mem) ((pthread_mutex_t *) (void*) (mem).data)
+
+/* While the locking-related APIs "exist" in later versions of
+ * Openssl, they tend to be placeholders or hardcoded values like:
+ * #define CRYPTO_get_locking_callback() (NULL)
+ * So the code will compile with strange warnings / errors like
+ * loop conditions always being false.
+ */
+#if !(S2N_OPENSSL_VERSION_AT_LEAST(1, 1, 0))
+
+static struct s2n_blob mutexes_mem = { 0 };
+static size_t mutexes_count = 0;
+
+static void s2n_locking_cb(int mode, int n, char *file, int line)
+{
+    pthread_mutex_t *mutexes = S2N_MUTEXES(mutexes_mem);
+    if (!mutexes_mem.data || n >= mutexes_count) {
+        return;
```

Comment:
For a bit I considered whether this should be fatal??? But dunno....

Reply:
"void" functions are tough.

While it would be great to avoid debugging the subtle problems if locking is a noop, shutting the entire process down for an error is pretty extreme. We don't do it anywhere else.